articles/mixin.dd

```diff
@@ -2,31 +2,130 @@ Ddoc
 
 $(D_S String Mixins,
 
-        $(P String Mixins (not to be confused with
+$(HEADERNAV_TOC)
+
+        $(P String mixins (not to be confused with
         $(DDLINK spec/template-mixin, Template Mixins, template mixins))
         enable string constants to be compiled as regular D code
         and inserted into the program.
-        Combining this with compile time manipulation of strings
-        enables the creation of domain-specific languages.
+        String mixins can be used in place of:)
+
+        $(UL
+        $(LI Declarations)
+        $(LI Statements)
+        $(LI Types)
+        $(LI Expressions)
         )
 
-        $(P For example, here we can create a template that generates
-        a struct with the named members:
+$(H2 $(LNAME2 declarations, Generating Declarations))
+
+        $(P A mixin can be used for a variable declaration:)
+---
+enum s = "int y;";
+mixin(s);
+y = 4;     // ok, mixin declared y
+---
+
+        $(P We can generate a declaration `struct Foo { int bar; }`
+        using string constants for the identifiers:
         )
 
 ---
-template GenStruct(string Name, string M1)
-{
-    const char[] GenStruct = "struct " ~ Name ~ "{ int " ~ M1 ~ "; }";
-}
+enum name = "Foo";
+enum m1 = "bar";
+mixin("struct " ~ name ~
+    "{ int " ~ m1 ~ "; }");
+
+enum f = Foo(2);
+static assert(f.bar == 2);
+---
+        $(P This can be encapsulated in an enum template for reuse:)
+---
+enum string genStruct(string name, string m1) =
+    "struct " ~ name ~ "{ int " ~ m1 ~ "; }";
+
+mixin(genStruct!("Foo", "bar"));
+---
+
+$(H2 $(LNAME2 multiple-args, Concatenation of Multiple Arguments))
 
-mixin(GenStruct!("Foo", "bar"));
+        $(P `mixin()` can also concatenate multiple arguments, converting any
+        non-strings to strings:)
 ---
-        $(P which generates:)
+enum name = "Foo";
+enum m1 = "bar";
+$(CODE_HIGHLIGHT enum val = 4;)
+mixin("struct ", name,
+    "{ int ", m1, " = ", $(CODE_HIGHLIGHT val), "; }");
+
+enum f = Foo();
+static assert(f.bar == 4);
 ---
-struct Foo { int bar; }
+
+$(H2 $(LNAME2 statements, Generating Statements))
+
+---
+int x = 3;
+mixin("
+    foreach (i; 0 .. 3)
+        writeln(x + i);
+    ");
+---
+        $(P This outputs:)
+$(CONSOLE
+3
+4
+5
+)
+        $(P See also: $(GLINK2 statement, MixinStatement).)
+
+$(H2 $(LNAME2 types, Generating Types))
+
+---
+mixin("int")* p;  // int* p
+mixin("int")[] a; // int[] a;
+mixin("int[]") b; // int[] b;
+---
+
+$(H2 $(LNAME2 expressions, Generating Expressions))
+
+---
+int foo(int x)
+{
+    return mixin("x +", 1) * 7;  // same as `return (x + 1) * 7`
+}
 ---
 
+$(H2 $(LNAME2 dsl, Domain Specific Languages))
+
+        $(P
+        Combining compile-time function execution with `mixin()`
+        enables the use of domain-specific languages:)
```

Comment:
Not a fan of this example. JavaScript is a general purpose programming language, a DSL is something like regex or SQL. Implementing `convertJS` would be a huge undertaking, and it doesn't reflect a real use case (or at least I don't see it). Why would you write this, instead of translating the JS offline and putting the result in your .d file?

I suggest using [pegged](https://github.com/PhilippeSigaud/Pegged) as an example.

Reply:
I don't see why any of that really matters. If the reader can't discern "real use cases" from a demonstrative example then they're not going to get anything from any example.

Besides, JavaScript did literally start as the ultimate domain specific language a long time ago.

Reply:
mixins are a confusing feature, and their working / purpose are [easily misunderstood](https://forum.dlang.org/post/hxoclgxkqatinlcnyzhr@forum.dlang.org). Consider the target audience for these articles: they're programmers who are new to D and might not have done any meta-programming beforehand. When you tell "`mixin("int y;")` is equivalent to `int y;`" the first question is "what does `mixin` do then?". 

The specification can be very dry in its explanations, but the article should help programmers understand what actual problems the feature solves. There's a tendency of new D users to complicate things with `mixin` when regular functions could be used, which is not helped by poor official examples.